columnflow/selection/cms/jets.py

```diff
@@ -53,61 +54,69 @@ def jet_veto_map(
 
     documentation: https://cms-jerc.web.cern.ch/Recommendations/#jet-veto-maps
     """
-
     jet = events.Jet
     muon = events.Muon[events.Muon.isPFcand]
 
+    # loose jet selection
     jet_mask = (
         (jet.pt > 15) &
-        (jet.jetId >= 2) &
+        (jet.jetId >= 2) &  # tight id
         (jet.chEmEF < 0.9) &
-        (ak.all(events.Jet.metric_table(muon) >= 0.2, axis=2))
+        ak.all(events.Jet.metric_table(muon) >= 0.2, axis=2)
     )
 
     # apply loose Jet puId in Run 2 to jets with pt below 50 GeV
-    if self.config_inst.campaign.x.year <= 2018:
-        jet_pu_mask = (events.Jet.puId >= 4) | (events.Jet.pt > 50)
+    if self.config_inst.campaign.x.run == 2:
+        jet_pu_mask = (events.Jet.puId >= 4) | (events.Jet.pt >= 50)
         jet_mask = jet_mask & jet_pu_mask
 
     # for some reason, math.pi is not included in the ranges, so we need to subtract a small number
     pi = math.pi - 1e-10
 
     # values outside [-pi, pi] are not included, so we need to wrap the phi values
-    jet_phi = ak.where(np.abs(events.Jet.phi) > pi, events.Jet.phi - 2 * pi * np.sign(events.Jet.phi), events.Jet.phi)
-
+    jet_phi = ak.where(
+        np.abs(events.Jet.phi) > pi,
+        events.Jet.phi - 2 * pi * np.sign(events.Jet.phi),
+        events.Jet.phi,
+    )
     variable_map = {
         "type": "jetvetomap",
-        "eta": jet.eta,
-        "phi": jet_phi,
+        "eta": jet.eta[jet_mask],
+        "phi": jet_phi[jet_mask],
     }
-
     inputs = [variable_map[inp.name] for inp in self.veto_map.inputs]
 
-    # apply the veto map
-    veto_map_result = ak.where(
-        jet_mask,
-        self.veto_map(*inputs),
-        -1,
-    )
+    # evalute the veto map only for selected jets
+    # (a map value of != 0 means the jet is vetoed)
+    veto_mask = jet_mask
+    flat_veto_mask = flat_np_view(veto_mask)
+    flat_veto_mask[flat_veto_mask] = ak.flatten(self.veto_map(*inputs) != 0)
 
-    # get the Jet veto mask (events containing such a jet should be vetoed)
-    veto_map_jet_mask = (veto_map_result > 0)
+    # store the per-jet veto mask
+    events = set_ak_column(events, "Jet.veto_map_mask", veto_mask)
```

Comment:
to me it looks like this is still just the `jet_mask`. I think you need to use the `layout_ak_array` function to transform the `flat_veto_mask` back into the correct shape.

Reply:
The `flat_veto_mask` is really just a view, so values changed in the view directly translate to changed values in the original, structured `veto_mask`.